utility/sourcemap_accessor.go

```diff
@@ -0,0 +1,192 @@
+package utility
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/go-sourcemap/sourcemap"
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+type SmapAccessor interface {
+	Fetch(smapId SmapID) (*sourcemap.Consumer, error)
+	RemoveFromCache(smapId SmapID)
+}
+
+type Enum string
+
+const (
+	InitError   Enum = "InitError"
+	AccessError Enum = "AccessError"
+	MapError    Enum = "MapError"
+	ParseError  Enum = "ParseError"
+)
+
+type SmapError struct {
+	Msg  string
+	Kind Enum
+}
+
+func (e SmapError) Error() string {
+	return fmt.Sprintf("%s", e.Msg)
+}
+
+type SmapID struct {
+	ServiceName    string
+	ServiceVersion string
+	Path           string
+}
+
+type SourcemapAccessor struct {
+	esClients []elasticsearch.Client
+	smapCache *lru.Cache
+}
+
+func NewSourcemapAccessor(esClients []elasticsearch.Client, cacheSize int) (*SourcemapAccessor, error) {
+	if esClients == nil || len(esClients) == 0 {
+		err := SmapError{Msg: "Sourcemap ESClient cannot be initialized.", Kind: InitError}
+		logp.Err(err.Error())
+		return nil, err
+	}
+
+	smapCache, err := lru.New(cacheSize)
+	if err != nil {
+		err := SmapError{Msg: fmt.Sprintf("Sourcemap cache cannot be initialized %s", err.Error()), Kind: InitError}
+		logp.Err(err.Error())
+		return nil, err
+	}
+	return &SourcemapAccessor{esClients: esClients, smapCache: smapCache}, nil
+}
+
+func (s *SourcemapAccessor) Fetch(smapId SmapID) (*sourcemap.Consumer, error) {
+	smapConsumer, err := s.fetchFromCache(smapId)
+	if err != nil {
+		logp.Err(fmt.Sprintf("Sourcemap fetching from Cache Error %s", err.Error()))
+	}
+	if smapConsumer == nil {
+		smapConsumer, err = s.fetchFromES(smapId)
+		if err != nil {
+			logp.Err(fmt.Sprintf("Sourcemap fetching from ES Error %s", err.Error()))
+			return nil, err
+		}
+		s.AddToCache(smapId, smapConsumer)
+	}
+	return smapConsumer, err
+}
+
+func (s *SourcemapAccessor) AddToCache(smapId SmapID, smap *sourcemap.Consumer) {
+	s.smapCache.Add(smapId.key(), smap)
```

Comment:
To Ron's [comment](https://github.com/elastic/apm-server/pull/381/files#r157163591) - we could track evictions using the return value of `Add` here.